cpp/src/parquet/geometry_util_internal.h

```diff
@@ -0,0 +1,395 @@
+        return "MULTILINESTRING";
+      case geometry_type::MULTIPOLYGON:
+        return "MULTIPOLYGON";
+      case geometry_type::GEOMETRYCOLLECTION:
+        return "GEOMETRYCOLLECTION";
+      default:
+        return "";
+    }
+  }
+};
+
+class WKBBuffer {
+ public:
+  WKBBuffer() : data_(NULLPTR), size_(0) {}
+  WKBBuffer(const uint8_t* data, int64_t size) : data_(data), size_(size) {}
+
+  void Init(const uint8_t* data, int64_t size) {
+    data_ = data;
+    size_ = size;
+  }
+
+  ::arrow::Result<uint8_t> ReadUInt8() {
+    if (size_ < 1) {
+      throw ParquetException("Can't read 1 byte from empty WKBBuffer");
+    }
+
+    size_ -= 1;
+    return *data_++;
+  }
+
+  ::arrow::Result<uint32_t> ReadUInt32(bool swap) {
+    if (size_ < sizeof(uint32_t)) {
+      return ::arrow::Status::SerializationError(
+          "Can't read 4 bytes from from WKBBuffer with ", size_, " remaining");
+    }
+
+    uint32_t value = ::arrow::util::SafeLoadAs<uint32_t>(data_);
+    data_ += sizeof(uint32_t);
+    size_ -= sizeof(uint32_t);
+    if (ARROW_PREDICT_FALSE(swap)) {
+      return value = ::arrow::bit_util::ByteSwap(value);
+    } else {
+      return value;
+    }
+  }
+
+  template <typename Coord, typename Func>
+  ::arrow::Status ReadDoubles(uint32_t n_coords, bool swap, Func&& func) {
+    if (n_coords == 0) {
+      return ::arrow::Status::OK();
+    }
+
+    size_t total_bytes = n_coords * sizeof(Coord);
+    if (size_ < total_bytes) {
+      return ::arrow::Status::SerializationError(
+          "Can't read ", total_bytes, " bytes from WKBBuffer with ", size_, " remaining");
+    }
+
+    if (ARROW_PREDICT_FALSE(swap)) {
+      Coord coord;
+      for (uint32_t i = 0; i < n_coords; i++) {
+        coord = ::arrow::util::SafeLoadAs<Coord>(data_);
+        for (uint32_t j = 0; j < coord.size(); j++) {
+          coord[j] = ::arrow::bit_util::ByteSwap(coord[j]);
+        }
+
+        func(coord);
+        data_ += sizeof(Coord);
+        size_ -= sizeof(Coord);
+      }
+    } else {
+      for (uint32_t i = 0; i < n_coords; i++) {
+        func(::arrow::util::SafeLoadAs<Coord>(data_));
+        data_ += sizeof(Coord);
+        size_ -= sizeof(Coord);
+      }
+    }
+
+    return ::arrow::Status::OK();
+  }
+
+  size_t size() { return size_; }
+
+ private:
+  const uint8_t* data_;
+  size_t size_;
+};
+
+struct BoundingBox {
+  BoundingBox(const std::array<double, 4>& mins, const std::array<double, 4>& maxes) {
+    std::memcpy(min, mins.data(), sizeof(min));
+    std::memcpy(max, maxes.data(), sizeof(max));
+  }
+  BoundingBox() : min{kInf, kInf, kInf, kInf}, max{-kInf, -kInf, -kInf, -kInf} {}
+
+  BoundingBox(const BoundingBox& other) = default;
+  BoundingBox& operator=(const BoundingBox&) = default;
+
+  void UpdateXY(std::array<double, 2> coord) { UpdateInternal(coord); }
+
+  void UpdateXYZ(std::array<double, 3> coord) { UpdateInternal(coord); }
+
+  void UpdateXYM(std::array<double, 3> coord) {
```

Comment:
it might be worth passing array > 3 byte reference (or more generally most of them by reference).  I guess without a benchmark it might be hard to tell.